lib/clAdapters/exchangeService/calendar/index.ts

```diff
@@ -0,0 +1,394 @@
+    userProfiles: UserProfile[] = [],
+    filterStartDate: Date,
+    filterEndDate: Date,
+    fields?: string
+  ) {
+
+    if (filterStartDate.getTime() > filterEndDate.getTime()) {
+      throw new Error(`filterStartDate must be less than or equal to filterEndDate`);
+    }
+
+    const { fieldNameMap } = ExchangeServiceCalendarAdapter;
+
+    const groupRunStats = {
+      success: true,
+      runDate: moment().utc().toDate(),
+      filterStartDate: filterStartDate,
+      filterEndDate: filterEndDate,
+      emails: userProfiles,
+      results: []
+    };
+
+    try {
+      const adapter = this;
+
+      // collect events for this group of emails
+      const results = await Promise.all(userProfiles.map(async(userProfile) => {
+        const individualRunStats = {
+          filterStartDate,
+          filterEndDate,
+          ...userProfile,
+          success: true,
+          runDate: moment().utc().toDate(),
+          errorMessage: null
+        };
+
+        this.setImpersonationUser(userProfile.emailAfterMapping);
+
+        try {
+          const result = await this.findItem(filterStartDate.toISOString(), filterEndDate.toISOString());
+          const items = _.get(result, 'ResponseMessages.FindItemResponseMessage.RootFolder.Items.CalendarItem');
+
+          const data = [];
+
+          if (items && items.length) {
+            for (const item of items) {
+              const out: { [key: string]: any } = {};
+
+              _.each(fieldNameMap, (have: string, want: string) => {
+                let modified = _.get(item, have);
+                if (/^dateTime/.test(want)) {
+                  modified = new Date(modified);
+                }
+                if (modified !== undefined) {
+                  out[want] = modified;
+                }
+              });
+
+              await this.attachAttendees(out, item);
+
+              out.attendees = _.map(out.attendees, (attendee: any) => {
+                const { Mailbox : email, ResponseType : response } = attendee;
+                return { address: email.EmailAddress, response };
+              });
+
+              _.remove(out.attendees, (attendee: any) => {
+                return !attendee.address;
+              });
+
+              out.canceled = adapter.parseBoolean(out.canceled);
+              out.allDay = adapter.parseBoolean(out.allDay);
+              out.hasAttachments = adapter.parseBoolean(out.hasAttachments);
+              data.push(out);
+            }
+          }
+
+          return Object.assign(individualRunStats, { data });
+        } catch (error) {
+          let errorMessage: any = error instanceof Error ? error : new Error(JSON.stringify(error));
+
+          if (/primary SMTP address must be specified/.test(errorMessage.message.toString())) {
+            errorMessage = {
+              type : 'InvalidPrimaryAddress',
+              error : new Error(`Email address: must use primary SMTP address for ${userProfile.emailAfterMapping}.`)
```

Comment:
Might be worth standardizing the error format between this and the Google errors (https://github.com/CrossLead/crosslead-adapters/blob/master/lib/clAdapters/google/errors.ts), otherwise the errors won't report under a specific grouping hash. CC @cynick.